README
/** \file
 * This file is used for the doxygen documentation. 
 */


/** \mainpage Documentation of lpzrobots
\par Organization:
	Max-Planck Institute for Mathematics in the Sciences\n	
	Inselstr. 22, 04103 Leipzig, Germany\n
	Dr. Georg Martius & Prof. Dr.  Ralf Der \n
        \n
	University Leipzig \n
	Institute for Computer Science  \n
	Dept. Intelligent Systems \n
	Frank Güttler \n
        Research Group: Technische Informatik\n


\par People:
Georg Martius, Ralf Der, Frank Hesse, Frank Güttler, Jörn Hoffmann\n
Former Contributors: Antonia Siegert, Marcel Kretschmann, Dominic Schneider, Claus Stadler
 
\section General

This is a collection of algorithms, simulations, and tools 
developed by the Robotics Group for Self-Organization of Control
(http://robot.informatik.uni-leipzig.de).\n
It consists of the following directories (click for details):
  - \ref selforg : controllers together with a small framework for using them,
	developed in the robotic group of Leipzig university
	yielding at self-organized behavior for various kinds of machines. 
  - \ref ode_robots : physics simulator based on ODE 
	(Open Dynamics Engine, see http://www.ode.org). 
	This includes such as robots, obstacles, utilities, 
         stuff for visualization with OSG 
        (OpenSceneGraph, see http://www.openscenegraph.org) and so on.
  - \ref guilogger : application that coordinates multiple gnuplot 
	windows and allows for an interactive display of data that is sent per pipe from another   - \ref matrixvix : application for interactive display of changing matrix and vector data
  - \ref configurator : a library implementing a GUI to change the parameters interactively
        which is otherwise done on the console
  - \ref ga_tools : genetic algorithms framework that can be used to together with 
        ode_robots or for independent simulations (not well maintained)
program 
  - \ref opende : directory with a snapshot of the open dynamics engine (release 0.11.1)
                  renamed to ode-dbl in order to avoid conflicts with packaged single
                  precision versions. It contains the capsule-box collision bugfix 
                  which is upstream (in svn)
                  (please follow the link for installation hints)

\section Inst Installation \& Startup

You have two different ways to get the simulator to work. 
- A) Install a package for your distribution.
     This is the quick 'n' easy way if you have root permissions,  see \ref InstA.
- B) Download the source tar ball and compile the simulator yourself, see \ref InstB

\subsection InstA Way A: Package installation
\subsubsection Ubuntu
  A package repository is hosted at
  https://launchpad.net/~georg-martius/+archive/lpzrobots \n
  Do on a terminal (or read the link: "read about installing" on that page):
\verbatim
sudo add-apt-repository ppa:georg-martius/lpzrobots
sudo apt-get update
\endverbatim
Then you can install <tt>lpzrobots</tt> as any other package do:
\verbatim
sudo apt-get install guilogger lpzrobots-oderobots
\endverbatim
This will automatically install all other dependencies.

That's it!

Now you can copy the sample simulations from
  <tt>/usr/share/lpzrobots/</tt> to your home directory and 
  continue reading in section \ref RunExample.

\subsubsection Other debian based systems (deb)
There is an install makefile located here: 
http://robot.informatik.uni-leipzig.de/software/packages/deb/install_deb_source.makefile .
Save it on you disk in an empty directory and run on the console
\verbatim
su -c apt-get install make
make -f install_deb_source.makefile
\endverbatim
That will download the newest packages, compile them and install it.
If something fails you can also redo parts of the process, see "make -f install_deb_source.makefile help".

The packages are: guilogger, matrixviz, ode-dbl, lpzrobots-selforg, and lpzrobots-oderobots
You can uninstall them later using the package manager (apt-get or synaptics)

\subsubsection Alternatively: Creating your own packages
You can do it step by step as follows:
Download all files from http://robot.informatik.uni-leipzig.de/software/packages/deb/current/ \n
Then do
\verbatim
# here you should not be root
 dpkg-source -x guilogger-0.*.dsc
 cd guilogger-0.* 
 dpkg-buildpackage -rfakeroot -b -uc
 cd ../
# become root (e.g. with su or sudo -s)
 dpkg -i guilogger-0.*.deb
\endverbatim
This you have to repeat for each package (e.g. replace guilogger by ode-dbl, lpzrobots-selforg, and lpzrobots-oderobots)

\subsection InstB Way B: Installation from source
\subsubsection Linux
 Check the <tt>Dependencies</tt> file for required packages.

\subsubsection MAC
 We recommend to install the ODE and OSG via macports (http://www.macports.org)
 (after installation of macports type on a console 
 <tt>sudo port install osg OpenSceneGraph</tt>) and see the <tt>Dependencies</tt> file.
 We will update the mac-related install procedures soon.

\subsubsection Instructions

- Download the source tar ball from 
  http://robot.informatik.uni-leipzig.de/software/current .
- Unpack file (<tt>tar -xvzf lpzrobots*.tar</tt>).
- Change into <tt> lpzrobots </tt> directory
- Call <tt> make help</tt> to get a help display and continue with
  <b><tt> make all</tt></b>. This will do everything. The first time you call it
  it will configure your build (you can reconfigure later with make conf).
  The following modules are compiled:
	- \ref matrixviz (not strictly necessary, only for displaying neuronal network parameters online) 
	- \ref guilogger (not strictly necessary, only for displaying parameters online (recommended)) 
	- \ref configurator library (not strictly necessary, only for changing parameters conveniently)
    	- Please note, that the make call will not fail if either of them failed to compile, because they are optional. You can type <tt> make guilogger </tt> and <tt> make matrixviz </tt> to compile them separately.
	- \ref selforg
	- \ref opende (our ode version with double precision) (required)
	- \ref ode_robots
	- \ref ga_tools 
  - if you do it step by step as display with <tt>make help</tt> 
     you need to use <tt>sudo make xxxx</tt> if the installation is into a system directory. (make all will do use it automatically if required). Note that the <tt>PATH</tt> variable needs to contain the PREFIX<tt>/bin</tt> (as checked by the configuration process). If you have multiple lpzrobots installations make sure the prefix for the current lpzrobots comes first in the <tt>PATH</tt> variable

- Now you are done with the installation and you can try a sample simulation see next section.

\subsection RunExample Run example Simulations
- Simulations are located in <tt>ode_robots/simulations/</tt>,<tt>ode_robots/examples/</tt>, and <tt>selforg/simulations/</tt>. This folders you can find in the lpzrobots tar files or if installed on your system under <tt>/usr/share/lpzrobots</tt>or<tt>/usr/local/share/lpzrobots</tt>. In the latter case copy the simulations to your home directory first.
- To start a simulation go into a simulation directory. 
- Call <tt>make</tt> to compile it.
- You can start the simulation by <tt>./start</tt>.
- For example when you want to start the template_sphererobot simulation type:
\verbatim
cd ode_robots/simulations/template_sphererobot
./start 
\endverbatim
- for optimization you can also use <tt>make opt</tt> which produces and <tt>start_opt</tt>. This is recommended to use after testing the code. 
- The following command line options are available 
  (type ./start -h for a full list of options): \n
\verbatim
Usage: ./start [-g [interval]] [-f [interval]] [-r seed] [-x WxH] [-fs]
                 [-pause] [-shadow N] [-noshadow] [-drawboundings] [-simtime [min]] [-threads N]
                 [-odethread] [-osgthread] [-savecfg]
        -g interval     use guilogger (default interval 1)
        -f interval     write logging file (default interval 5)
        -m interval     use matrixviz (default interval 10)
        -s "-disc|ampl|freq val"        use soundMan
        -r seed         random number seed
        -x WxH          * window size of width(W) x height(H) is used (default 640x480)
        -fs             fullscreen mode
        -pause          start in pause mode
        -nographics             start without any graphics
        -noshadow       disables shadows and shaders (same as -shadow 0)
        -shadow [0..5]]         * sets the type of the shadow to be used
                        0: no shadow, 1: ShadowVolume, 2: ShadowTextue, 3: ParallelSplitShadowMap
                        4: SoftShadowMap, 5: ShadowMap (default)
        -shadowsize size        * sets the size of the shadow texture (default 2048)
        -drawboundings  enables the drawing of the bounding shapes of the meshes
        -simtime min    limited simulation time in minutes
        -savecfg        safe the configuration file with the values given by the cmd line
        -threads N      number of threads to use (default is the number of processors)
        -odethread      * if given the ODE runs in its own thread. -> Sensors are delayed by 1
        -osgthread      * if given the OSG runs in its own thread (recommended)
        * this parameter can be set in the configuration file ~/.lpzrobots/ode_robots.cfg
\endverbatim
- On some machine the program crashed right away because of graphic card incompatibilities,
  try: <tt>./start -noshadow</tt>
- Have a look at the console after starting the program, 
  there you will find some further information for the usage of the program.
- E. g. with Ctrl+C (on the terminal) you get a interactive console interface 
  which can be used to modify  parameters on the fly.
- For starting your own simulation see paragraph "How to Start Your Own Simulation" 
  in \ref ode_robots.
- For a well documented examples of a main.cpp of a simulation and a robot .cpp file click 
  the tab "Examples" at the top of this page.


\section Documentation

- This manual can be found at http://robot.informatik.uni-leipzig.de/software
- More information on the used self-organization algorithm is available at  
  http://robot.informatik.uni-leipzig.de/research
- The original ODE documentation can be found at http://opende.sourceforge.net/wiki/index.php/Main_Page
- The OSG documentation can be found at http://www.openscenegraph.org/projects/osg/wiki/Support

*/


ChangeLog
2016-01-19 martius (0.8.1)
	fixing some small bugs
	motorbabbler more configureable

2014-06-06 martius (0.8)
	use of C++11 (need gcc 4.7 or higher)
	Robot interface for ode_robots changed, see UPDATE
	generic sensor and motors can not be used widely, (Servo motors as well)
	Sensor and motor descriptions routed from robots to controllers
	Generic set of config variables from console -set "{key1=val1 key2=val2 ...}"
	Ctrl-h resets robots to initial pose (of main primitive)
	Fixation and displacement of robots integrated
	video recording also records logfiles and saves agents automatically
	encodevideo.sh creates a src archive with sources and logfiles
	PlotOptions allow filtering. From commandline e.g. -g 10 "{+x,+y}" or "{-C}"
	Bugfixes: Following camera mode, irsensor-nographics-crash,...
	RelatvePositionSensor fixed
	TorqueSensors added
	Contacts can be drawn with -drawcontacts
	commandline inputs go into logfiles
	Matrix uses move-assignment from C++11, speedup

2011-11-11 martius (0.7)
	Build system: use config scripts like selforg-config etc.
	Configurator added (some stability issues remain)
	Operators/Manipulators: observe agent and can turn it over of lift up
	  see LimitOrientationOperator and LiftUpOperator
	Colors and ColorSchema: named colors and alias sets
   	 color palette files (gimp format)
	 alias files: simple tab seperated file


2010-07-05 martius (0.65)
	Build system changed. Object files come to separate directory.
	M4 Makefile for simulations
	Angularmotors bug fixed. Axis were global.
	New Servos (XAxisVel) that control internally the velocity work great now


2010-03-25 martius (0.6)
	long time no update in the Changelog file. Almost everthing changed ;-)
	Main points
	1. Camera Support
	2. Normal target is with -O1 and there is an additional debug target


2008-04-17 martius
	1. OSG2 port
	New viewer is used.
	Various shadow algorithms are avaiable
	2. Multithreading build in using QuickMP
	Controllers (Agent::step) run in parallel
	Threadnumber is the number of processors.
	ODE can also run in a separate thread if wanted, this has some consequences:
	  - sensors are one timestep delayed
	OSG can also run in a separate thread if wanted, this has some consequences:
	  - sometimes the program just stalls right after the start which is due
	    to some problems with the Xlib and concurrent access
	3. Random generator with state
	Agents have there own random generator and
	pass it to controller and wiring
	4. Configuration file
	a config file (~/.lpzrobots/ode_robots.cfg) loads defaults for
	 e.g. shadow, window size, threads and so on.
	The cfg file is written at the first start,
	 and it can be also saved with -savecfg flag
	5. Support for simple sound signals
	very simple sound sensor and sound emitting with graphic represenation
	6. Binary distribution creation script


2007-03-26 martius
	1. new collision treatment
	Class substance carries information about material of each primitive.
	The substance for new primitives is taken from odeHandle.
	For an example see simulations/tests.
	On collision the substances of two colliding primitives are combines to
	contact surface parameter.
	There is a global list of ignored pairs of geoms. If you create a joint the
	two parts are automatically inserted in this list. If you want to specify
	additional pairs, which should not collide, then use odeHandle.addIgnoredPair().
	If you want to ignore an entire space then use odeHandle.addIgnoredSpace().
	In order to stay backwards compatible each robot is asked for the collision
	handling as before. If no robots handles the collision the general collision
	handling is used. See also simulations/tests.

	2.There is a new commandline interface using gnu readline.
	Features: Commands
	param=val      sets PARAM of all objects to VAL.
	help           Display this text.
	?              Synonym for `help'.
	list           Lists all objects and agents.
	ls             Synonym for `list'.
	set            syntax: set [OBJECTID] PARAM=VAL; sets parameter of Object (or all all objects) to value.
	store          Stores controller of AGENTID to FILE.
	load           Loads controller of AGENTID from FILE.
	show           [OBJECTID]: Lists paramters of OBJECTID or of all objects (if no id given).
	view           Synonym for `show'.
	quit           Quit program.

	There is also a history and all features you know from you terminal console e.g.:
	Up / Down  previous and next in history
	TAB        auto completion
	Ctrl+r     reverse search in history
	...

	With store and load you can save the controller of any agent

2006-12-12 martius
	1. agents receive comments about changes in configuration values
	from simulations.cpp


2006-12-11 martius
	1. controllinterval wie schon angekuendigt ist jetzt die richtige
	Verhalten bezueglich des controllintervals eingecheckt. D.h. in
	jedem Simulationsschritt werden die Motorsignale an den Roboter
	geschickt, aber nur aller "controllinterval" Schritte werden sie
	Sensoren abgefragt und der Controller benutzt.

	2. derivativewiring
	2a: Reihenfolge hat sich geaendert: Jetzt: sensoren einer Qualitaet in
	Bloecken. Also erst alle Sensorwerte (Id), dann alle ersten Ableitungen
	(FirstD), dann alle zweiten Ableitungen (SecondD)
	2b: Rauschen: Nur der erste Block an Sensorwerten wird verrauscht. Also wenn
	man Id und FirstD benutzt, dann sind nur Id verrauscht. Wenn man FirstD und
	SecondD benutzt, dann ist nur FirstD verrauscht.
	2c: Defaults: es gibt jetzt zwei Default Funktionen:
		getDefautConf(): Id und FirstD, kein Smoothing, kein Scaling
		getDetaultConf1() Nur FirstD, Smoothing ueber 2 Zeitschritte, Scaling mit 5
		(die alten Einstellungen: Smoothing ueber 20 Zeitschritte, Scaling mit 10)
	2d: bei blind motors wird jetzt direkt die Nummer angegeben.

	3. agent hat eigenen noisefactor:
	beim Konstuktor eines Agents, kann man jetzt einen noisefactor angeben.
	z.B. OdeAgent* agent  = new OdeAgent(plotoptions, 0.1);

ecbrobots/ecb.cpp
/***************************************************************************
 *   Copyright (C) 2005 by Robot Group Leipzig                             *
 *    martius@informatik.uni-leipzig.de                                    *
 *    fhesse@informatik.uni-leipzig.de                                     *
 *    der@informatik.uni-leipzig.de                                        *
 *    guettler@informatik.uni-leipzig.de                                   *
 *    marcoeckert@web.de                                                   *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 *                                                                         *
 *   $Log$
 *   Revision 1.18  2011-03-22 16:36:29  guettler
 *   - ECB is now inspectable
 *
 *   Revision 1.17  2011/02/04 12:59:05  wrabe
 *   - convert function for short values added
 *
 *   Revision 1.16  2011/01/24 14:16:25  guettler
 *   - removed deprecated comments
 *   - setMotors now uses currentNumberMotors instead of maximum value
 *
 *   Revision 1.15  2010/11/26 12:22:37  guettler
 *   - Configurable interface now allows to set bounds of paramval and paramint
 *     * setting bounds for paramval and paramint is highly recommended (for QConfigurable (Qt GUI).
 *   - bugfixes
 *   - current development state of QConfigurable (Qt GUI)
 *
 *   Revision 1.14  2010/11/11 15:34:59  wrabe
 *   - some extensions for QMessageClient (e.g. quitServer())
 *   - fixed some includes
 *
 *   Revision 1.13  2010/11/10 09:32:00  guettler
 *   - port to Qt part 1
 *
 *   Revision 1.12  2009/08/18 14:49:37  guettler
 *   implemented COMMAND_MOTOR_MAX_CURRENT
 *
 *   Revision 1.11  2009/08/11 15:49:05  guettler
 *   Current development state:
 *   - Support of communication protocols for XBee Series 1, XBee Series 2 and cable mode
 *   - merged code base from ecb_robots and Wolgang Rabes communication handling;
 *     ECBCommunicator (almost) entirely rewritten: Use of Mediator (MediatorCollegues: ECB),
 *     Callbackble (BackCaller: SerialPortThread)
 *   - New CThread for easy dealing with threads (is using pthreads)
 *   - New TimerThreads for timed event handling
 *   - SerialPortThread now replaces the cserialthread
 *   - QGlobalData, ECBCommunicator is now configurable
 *   - ECBAgent rewritten: new PlotOptionEngine support, adapted to new WiredController structure
 *   - ECBRobot is now Inspectables (uses new infoLines functionality)
 *   - ECB now supports dnsNames and new communication protocol via Mediator
 *   - Better describing command definitions
 *   - SphericalRobotECB: adapted to new ECB structure, to be tested
 *
 *   Revision 1.10  2009/03/25 11:06:55  robot1
 *   updated version
 *
 *   Revision 1.9  2008/08/15 13:16:58  robot1
 *   add PORT PIN configuration for ECBs
 *
 *   Revision 1.8  2008/08/12 11:45:20  guettler
 *   plug and play update, added some features for the ECBRobotGUI
 *
 *   Revision 1.7  2008/07/16 14:37:17  robot1
 *   -simple getc included
 *   -extended config on reset
 *   -minor changes
 *
 *   Revision 1.6  2008/07/16 07:38:42  robot1
 *   some major improvements
 *
 *   Revision 1.5  2008/04/11 06:31:16  guettler
 *   Included all classes of ecbrobots into the namespace lpzrobots
 *
 *   Revision 1.4  2008/04/11 06:15:48  guettler
 *   Inserted convertion from byte to double and backwards for motor and sensor values
 *
 *   Revision 1.3  2008/04/08 10:11:03  guettler
 *   alpha testing
 *
 *   Revision 1.2  2008/04/08 09:09:09  martius
 *   fixed globaldata to pointer in classes
 *
 *   Revision 1.1.1.1  2008/04/08 08:14:30  guettler
 *   new ecbrobots module!
 *
 *                                                                         *
 ***************************************************************************/
#include "ecb.h"
#include "commanddefs.h"
#include "QECBCommunicator.h"
#include "ECBCommunicationData.h"
#include "QGlobalData.h"

#include <iostream>
#include <sstream>
#include <assert.h>

using namespace std;

namespace lpzrobots {
  ECB::ECB(QString dnsName, QGlobalData& globalData, ECBConfig& ecbConfig) :
    Configurable(dnsName.toStdString(), "$ID$"), Inspectable(dnsName.toStdString()), MediatorCollegue(globalData.comm), globalData(&globalData), ecbConfig(ecbConfig), dnsName(dnsName) {
    failureCounter = 0;
    initialised = false;
  }

  ECB::~ECB() {
  }
  
  void ECB::sendMotorValuesPackage() {
    // at first start of ECB, it must be initialized with a reset-command
    assert(initialised && failureCounter <= globalData->maxFailures);

    globalData->textLog("ECB(" + dnsName + "): sendMotorPackage()!");

    // prepare the communication-protocol
    ECBCommunicationEvent* event = new ECBCommunicationEvent(ECBCommunicationEvent::EVENT_REQUEST_SEND_COMMAND_PACKAGE);

    event->commPackage.command = COMMAND_MOTORS;
    event->commPackage.dataLength = currentNumberMotors;
    // set motor-data
    int i = 0;
    // motorList was update by ECBAgent->ECBRobot:setMotors()->(to all ECBs)-> ECB:setMotors()
    FOREACH (list<motor>,motorList,m) {
      // Agent and Controller process with double-values
      // The ECB(hardware) has to work with byte-values
      event->commPackage.data[i++] = convertToByte((*m));
    }
    informMediator(event);
  }
  
  void ECB::sendMaxMotorCurrent(uint8 maxCurrent, uint8 motorboardIndex) {
    // at first start of ECB, it must be initialized with a reset-command
    assert(initialised && failureCounter <= globalData->maxFailures);

    globalData->textLog("ECB(" + dnsName + "): sendMaxMotorCurrent(" + maxCurrent + ", " + motorboardIndex + ")!");

    // prepare the communication-protocol
    ECBCommunicationEvent* event = new ECBCommunicationEvent(ECBCommunicationEvent::EVENT_REQUEST_SEND_COMMAND_PACKAGE);

    event->commPackage.command = COMMAND_MOTOR_CURRENT_LIMIT;
    event->commPackage.dataLength = 2;

    event->commPackage.data[0] = motorboardIndex;
    event->commPackage.data[1] = maxCurrent;
    informMediator(event);
  }

  /**
   * Send reset command to the ecb and
   * receive the number of motors and sensors
   * @return
   */
  void ECB::sendResetECB() {

    //if (globalData->debugOutput)
    globalData->textLog("ECB: resetECB!(" + dnsName + ")");

    // globalData->comm->flushInputBuffer();

    initialised = false;

    // TODO: send specific config settings to ECB!

    generateAndFireEventForCommand(COMMAND_RESET);
  }

  void ECB::makeBeepECB() {
    generateAndFireEventForCommand(COMMAND_BEEP);
  }
  
  void ECB::stopMotors() {
    generateAndFireEventForCommand(COMMAND_MOTOR_STOP);
  }
  
  void setMaxMotorCurrent(uint8 maxCurrent, uint8 motorboardIndex = 0) {

  }

  void ECB::startMotors() {
    generateAndFireEventForCommand(COMMAND_MOTOR_START);
  }
  
  /**
   * The ECBRobot holds a list of motor-values. To distribute this values
   * to all connected ECBs is done by following: Every ECB get the hole motor-value-array.
   * Additional the beginIndex for start-position (first ECB have the start-pos.=0).
   * The ECB-Function setMotors will add the values into motor-list itself, until the
   * number of his motorNumber. After that, this function will return the number of
   * added values. This number is equal to the start-position of the next ECB motorList.
   *
   * motorArray - hold the new motor-values from controller
   * beginIndex - the start-pos of motor-array that will be add into ECB-motorList
   * maxIndex   - size of motorArray
   *
   * return     - the number of adding motor-values
   */
  int ECB::setMotors(const motor* motorArray, int beginIndex, int maxIndex) {
    motorList.clear();
    int i = 0;
    for (i = beginIndex; i < currentNumberMotors; i++) {
      motorList.push_back(motorArray[i]);
      if (i == maxIndex)
        break;
    }
    return (i - beginIndex);
  }
  
  int ECB::getNumberMotors() {
    return (currentNumberMotors > ecbConfig.maxNumberMotors ? ecbConfig.maxNumberMotors : currentNumberMotors);
  }
  
  int ECB::getNumberSensors() {
    return (currentNumberSensors > ecbConfig.maxNumberSensors ? ecbConfig.maxNumberSensors : currentNumberSensors);
  }
  
  int ECB::getMaxNumberMotors() {
    return (ecbConfig.maxNumberMotors);
  }
  
  int ECB::getMaxNumberSensors() {
    return (ecbConfig.maxNumberSensors);
  }

  /// helper functions

  double ECB::convertToDouble(uint8 byteVal) {
    // 0 <= byteVal <= 255
    if (byteVal >= 255)
      byteVal = 255;
    if (byteVal <= 0)
      byteVal = 0;

    //globalData->textLog("Converted" + (unsigned int)byteVal + "to doubleval = " + (((double) (byteVal - 128)) / 128.));
    return (((double) (byteVal - 128)) / 128.);

  }

  uint8 ECB::convertToByte(double doubleVal) {
    // insert check byteVal<255
    // -1 <= doubleVal <= 1
    if (doubleVal >= 1)
      doubleVal = 1;
    if (doubleVal <= -1)
      doubleVal = -1;

    int byteVal = (int) ((doubleVal + 1.) * 128.0);
    if (byteVal > 255)
      byteVal = 255;
    if (byteVal < 0)
      byteVal = 0;
    //globalData->textLog("Converted" + doubleVal + "to byteval = " + (unsigned int)byteVal);
    return (uint8) byteVal;
  }

  double ECB::convertToDouble(short shortVal, short minBound, short maxBound) {
    assert(minBound <= maxBound);
    assert(minBound !=0 && maxBound!=0);
    if (shortVal > maxBound)
      shortVal = maxBound;
    if (shortVal < minBound)
      shortVal = minBound;
    // shift zero point!
    shortVal += (maxBound + minBound)/2;
    if (shortVal > 0) {
      return (shortVal / (double)maxBound);
    } else if (shortVal < 0) {
      return -(shortVal / (double)minBound);
    } else {
      return 0;
    }
  }

  short ECB::convertToShort(double doubleVal, short minBound, short maxBound) {
    if (doubleVal >= 1)
      doubleVal = 1;
    if (doubleVal <= -1)
      doubleVal = -1;

    int zero = (maxBound + minBound)/2;
    int range = maxBound - minBound;
    short shortVal = (short) (doubleVal * 0.5 * range + zero);
    return shortVal;
  }

  /// STORABLE INTERFACE

  /** stores the object to the given file stream (binary). */
  bool ECB::store(FILE* f) const {
    // viel Spaß!
    return true;
  }
  
  /** loads the object from the given file stream (binary). */
  bool ECB::restore(FILE* f) {
    // xml parser anwerfen
    // viel Spaß!
    return true;
  }

  std::string ECB::getChannelDescription() {
    if (!initialised)
      return "";

    /*stringstream ss;
     std::string s="";
     for (int i=0;i<ecbConfig.md23_sensors;i++) {
     ss + "mot(" + address + ")";
     }
     for (int i=0;i<ecbConfig.pcf_sensors;i++) {
     ss + "pcf(" + address + ")";
     }
     for (int i=0;i<ecbConfig.adc_sensors;i++) {
     ss + "adc(" + address + ")";
     }
     s = ss.str(); */

    globalData->textLog("ECB: getChannelDescription():[" + QString(descriptionLine.c_str()) + "]");

    return descriptionLine;
  }

  void ECB::commandDimensionReceived(ECBCommunicationEvent* event) {
    globalData->textLog("dimension package received: ");
    /*********************************************************************
     // GET the new description of connected motors
     // and sensors at ECB(hardware)
     // Each sensor-name should be unique to identify the sensors as well
     // The names are separated by a space-sign
     *********************************************************************/
    ECBCommunicationData result = event->commPackage;

    // set number of motors and sensors that are in use (current)
    currentNumberMotors = result.data[0];
    currentNumberSensors = result.data[1];

    globalData->textLog("[" + dnsName + "] found motors: " + QString::number(currentNumberMotors) + ", sensors: " + QString::number(currentNumberSensors));

    // modify the data-values with the ECB-DnsName
    stringstream ss;
    QString description;
    globalData->textLog("[" + dnsName + "] descriptionLine:");

    for (int i = 2; i < result.dataLength; i++) {
      if (result.data[i] == ' ') {
        ss << "(" << dnsName.toStdString() << ")";
        globalData->textLog("   - " + description);
        description.clear();
      } else {
        ss << (uchar) result.data[i];
        description.append(result.data[i]);
      }
    }
    // add the last dnsName to stringstream if its not empty
    if (result.dataLength > 0)
      ss << "(" + dnsName.toStdString() << ")";

    if (currentNumberSensors < ecbConfig.maxNumberSensors) {
      for (int i = currentNumberSensors; i < ecbConfig.maxNumberSensors; i++) {
        ss << " -";
      }
    }

    // complete description as a string-line
    descriptionLine = ss.str();

    if (currentNumberMotors > ecbConfig.maxNumberMotors) {
      globalData->textLog("Warning: ECB " + dnsName + " reported more motors than permitted and configured respectively!");
    }

    if (currentNumberSensors > ecbConfig.maxNumberSensors) {
      globalData->textLog("Warning: ECB " + dnsName + " reported more sensors than permitted and configured respectively!");
    }

    //globalData->textLog("[" + dnsName + "] descriptionLine:[" + QString(descriptionLine.c_str()) + "]");

    initialised = true;
    failureCounter = 0;

    // set max motor current if deviating from default value
    if (ecbConfig.maxMotorCurrent != DEFAULT_MAX_MOTOR_CURRENT)
      sendMaxMotorCurrent(ecbConfig.maxMotorCurrent, 0);
  }

  void ECB::commandSensorsReceived(ECBCommunicationEvent* event) {
    globalData->textLog("sensor package received");
    ECBCommunicationData result = event->commPackage;

    /// fill sensorList which will be the input for agent or controller
    sensorList.clear();
    for (int i = 0; i < result.dataLength; i++) {
      sensorList.push_back(convertToDouble(result.data[i]));
    }

    // reset the counter, because communication was successful
    failureCounter = 0;
  }

  /**
   * Is called when the mediator informs this collegue that an event
   * has to be performed by this collegue instance.
   */
  void ECB::doOnMediatorCallBack(MediatorEvent* event) {
    ECBCommunicationEvent* commEvent = static_cast<ECBCommunicationEvent*> (event);
    switch (commEvent->type) {
      case ECBCommunicationEvent::EVENT_PACKAGE_SENSORS_RECEIVED:
        commandSensorsReceived(commEvent);
        break;
      case ECBCommunicationEvent::EVENT_PACKAGE_DIMENSION_RECEIVED:
        commandDimensionReceived(commEvent);
        break;
      case ECBCommunicationEvent::EVENT_REQUEST_SEND_MOTOR_PACKAGE:
        if (initialised)
          sendMotorValuesPackage();
        else
          sendResetECB();
        break;
      case ECBCommunicationEvent::EVENT_REQUEST_SEND_MOTOR_STOP_PACKAGE:
        if (initialised)
          stopMotors();
        break;
      case ECBCommunicationEvent::EVENT_COMMUNICATION_ANSWER_TIMEOUT:
        globalData->textLog("ECB(" + dnsName + ") did not answer: ");
        if (initialised) {
          if (failureCounter >= globalData->maxFailures) { // try to send reset next time
            globalData->textLog(" failure count=" + QString::number(failureCounter + 1) + " reached maximum, reset to initial state.");
            initialised = false;
            failureCounter = 0;
          } else {
            failureCounter++;
            globalData->textLog(" failure count=" + failureCounter);
          }
        } else
          // do nothing, try later again
          globalData->textLog(" no initial response to reset command.");
        break;
      default:
        break;
    }

  }

  void ECB::generateAndFireEventForCommand(uint8 command) {
    ECBCommunicationEvent* event = new ECBCommunicationEvent(ECBCommunicationEvent::EVENT_REQUEST_SEND_COMMAND_PACKAGE);

    event->commPackage.command = command;
    event->commPackage.dataLength = 0;

    informMediator(event);
  }

} // end namespace lpzrobots
